### PR TITLE
fix(cursor): namespace bare client tools on the wire

### PR DESCRIPTION
**Pull request by HAEGONG**
## Summary

- Fix Claude Desktop requests through the Cursor provider that could fail with an empty completion/HTTP 502 when client-owned bare tool names collided with Cursor's private tool namespace. The same account and model worked through Codex's Responses path, showing this was a tool-mapping interoperability bug rather than a subscription-entitlement failure.
- Namespace bare client-owned Cursor tools as `ocx_client_<name>` on the provider wire to avoid collisions with Cursor's private bare-tool namespace.
- Preserve existing wire names for proxy-owned tools (`exec`, `wait`, `exec_command`, `shell_command`, `apply_patch`, `edit_file`, `multi_edit`, and `tool_search`) and map aliased calls back to their original client names before emitting responses.
- Reuse the same provider-isolated alias for structured replay, with focused regression coverage for declarations, call mapping, freeform validation, stream completion, and both Cursor transports.
- Non-blocking P2 follow-up: text-only external replay continues to display the semantic client tool name while native structured replay uses the wire alias; `composer-2.5` can exercise both replay shapes. This patch intentionally leaves that presentation-level consistency question out of scope because call routing already uses the mapped identity.

## Verification

- `bun run prepush` (through the installed pre-push hook) — passed, including typecheck, the full test suite (`16,350` pass, `12` skip, `0` fail, plus every serial suite), and the privacy scan.
- `bun test tests/update-stop-first.test.ts` — `15` pass, `0` fail.
- `bun test tests/shutdown-launcher.test.ts` — `3` pass, `0` fail.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults.

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Cursor tool-call replay and completion handling for client-provided tools.
  * Prevented naming conflicts by consistently isolating client tools with an `ocx_client_` prefix.
  * Improved handling of incomplete or malformed freeform tool arguments.
  * Preserved correct tool identities and results across live connections, reconnects, and clean connection endings.

* **Tests**
  * Expanded coverage for tool discovery, replay, streaming events, and transport edge cases.

### CODE REVIEW

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 66 / 80

이 PR은 Cursor 어댑터가 클라이언트가 보낸 네임스페이스 없는(bare) 도구를 프로바이더 쪽으로 광고할 때, Cursor 쪽 비공개 bare 도구 이름과 겹치지 않도록 ocx_client_ 접두사를 붙이는 수정이다. 지금 dev(870a2adb6, #3013 이후)의 src/adapters/cursor/tool-definitions.ts에 있는 cursorToolWireName은 그냥 namespacedToolName만 돌려준다. 그래서 클라이언트가 read, Glob, script처럼 bare 이름을 주면, 그 이름이 Cursor 내부 도구 공간과 충돌할 수 있다. 이 PR은 그 충돌을 피하려고 클라이언트 소유 bare 도구만 와이어에서 ocx_client_<이름>으로 바꾸고, 응답이 돌아올 때는 다시 원래 클라이언트 이름으로 되돌린다.

프록시가 직접 소유하는 bare 도구(exec, wait, exec_command, shell_command, apply_patch, edit_file, multi_edit, tool_search)는 접두사를 붙이지 않는다. 이미 네임스페이스가 있는 도구도 그대로다. 즉 클라이언트가 가져온 bare 이름만 격리한다. src/adapters/cursor/live-transport.ts는 이미 cursorToolWireName(tool)로 cursorToolNameMap 키를 만들고 있으니, cursorToolWireName만 바꾸면 맵·스키마·광고 이름이 같이 따라간다. protobuf-request.ts의 구조화 리플레이도 같은 와이어 이름을 쓰도록 맞춰 두었다.

테스트는 선언, 호출 매핑, freeform 인자 검증, 스트림 완료, live/HTTP1 양쪽을 건드린다. cursor-protobuf-events.test.ts에 alias에서 클라이언트 이름으로 복원, 잘못된 freeform 거부, partial freeform 유지 케이스가 생겼다. cursor-tool-definitions.test.ts는 프록시 소유 목록이 그대로 남는지, 안내 노트에 와이어 이름(ocx_client_Glob 등)이 보이는지도 잠가 두었다. 작성자가 말한 대로 bun run prepush와 전체 스위트가 로컬에서 통과했다고 한다. CI의 enforce-target/hygiene/label도 통과 상태다.

지금 dev가 최적화하는 축(#3013 릴리스 버전 라인, Kiro terminal completion, 대시보드 사이드카)과는 다른 스택이지만, Cursor 클라이언트 도구 경로는 여전히 핵심 런타임이다. 타입/설정 분할 캠페인에 걸려 무효화될 종류의 PR도 아니다. 스코프도 좁고 review-ready 라벨까지 붙어 있다.

라인 330 - cursorToolWireName에 name이 이미 ocx_client_read인 경우 ocx_client_ocx_client_read가 되는 동작을 테스트가 고정한다. 이미 접두사가 붙은 이름을 클라이언트가 그대로 넘기면 이중 접두사가 난다. 의도된 방어인지, 입력을 normalize해서 한 번만 붙일지 한 줄 정책이 필요하다.
경로/심볼 cursorToolWireName·안내 노트 - 모델에게 보이는 도구 이름이 ocx_client_Glob처럼 바뀐다. 호출 라우팅은 맵으로 되돌리지만, 사용자가 UI/로그에서 보는 이름과 모델이 고르는 이름이 어긋날 수 있다. 작성자도 text-only 리플레이 vs 구조화 리플레이 표시 차이를 P2로 남겨 두었다.
경로/심볼 CURSOR_PROXY_OWNED_BARE_TOOL_NAMES - tool_search를 상수로 올려 프록시 소유 집합에 넣었다. 이후 bare 프록시 도구가 하나 더 생기면 이 Set을 깜빡 잊기 쉽다. 목록의 단일 출처가 있는지 확인하면 좋다.
경로/심볼 freeformToolNames - live-transport는 freeform 집합을 여전히 클라이언트 semantic 이름(script)으로 두고, clientToolNames/맵은 와이어 이름(ocx_client_script)을 쓴다. 테스트로 잠가 두었지만, 이 이분법이 다른 이벤트 경로에도 빠짐없이 적용되는지 한 번 더 훑을 가치가 있다.

메인테이너의 판단이 필요한 지점
- 이중 접두사(ocx_client_ocx_client_*)를 허용할지, 입력에서 접두사를 벗겨 한 번만 붙일지
- text-only 리플레이에 semantic 이름이 남는 P2를 이번 PR에 묶을지, 후속으로 열지
- Cursor 쪽 비공개 bare 네임스페이스와 실제로 충돌했던 재현(어느 클라이언트 도구 이름인지)을 이슈/본문에 한 줄 더 남길지

너의 추천
dev에 리베이스된 상태와 로컬 prepush 통과를 신뢰하면, 이중 접두사 정책만 코멘트로 확정한 뒤 머지해도 된다. P2 표시 차이는 별 이슈로 열어 두고 이번 PR 스코프는 유지하는 쪽을 추천한다.

이 댓글은 grok-bot이 작성했습니다

File: src/adapters/cursor/tool-definitions.ts [hunk 1]
```diff
@@ -320,10 +321,39 @@ export function cursorRequestAdvertisesStructuredEdits(
...
+function isCursorBareClientToolWireAliased(
+  tool: Pick<OcxTool, "namespace" | "name">,
+): boolean {
+  return !tool.namespace
+    && !CURSOR_PROXY_OWNED_BARE_TOOL_NAMES.has(tool.name);
+}
+
 export function cursorToolWireName(tool: Pick<OcxTool, "namespace" | "name">): string {
+  if (isCursorBareClientToolWireAliased(tool)) {
+    return `${CURSOR_CLIENT_TOOL_WIRE_PREFIX}${tool.name}`;
```
[the remark this answers is not in the document]
Comment (HAEGONG) [reply src/adapters/cursor/tool-definitions.ts:348]: Fixed in `eca5ff72c`. Cursor-generated `ocx_client_*` wire aliases are now fallback-only when the catalog already contains a tool with that exact semantic name. This preserves the lone-`read` wire-alias fallback while ensuring an explicit `ocx_client_read` selector advertises only the semantically named tool. Added regression coverage for both named `tool_choice` and `allowedTools`. The installed pre-push hook passed at this head: typecheck, full suite (`16,350` pass / `12` skip / `0` fail, plus serial suites), and privacy scan.

**Review by Ingwannu — APPROVED** [review]
Approved on exact head eca5ff72c8dde21e1ad6e5ee57b5b868028cb7ab. The patch isolates ordinary client-owned bare names on the Cursor wire while preserving proxy-owned and already-namespaced identities, maps calls back to the semantic client name, and reuses the same alias for structured replay. The explicit semantic-name precedence case, freeform validation, clean-EOF handling, and both Cursor transports are covered. Exact-head CI is fully green and diff check is clean. The documented text-only versus structured replay label difference can remain a non-blocking P2 follow-up; do not widen this patch into presentation cleanup.
